### PR TITLE
feat: per wallet lock

### PR DESCRIPTION
**Pull request by r4mmer**
### Acceptance Criteria

- `SENT_TX` lock should be applied on a per-wallet basis

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/lock.js [hunk 1]
```diff
@@ -68,4 +69,42 @@ class Lock {
   }
 }
 
-export const lock = new Lock();
+class GlobalLock {
+  constructor() {
+    this.wallets = {};
```
Comment (pedroferreira1) [inline src/lock.js:74]: suggestion (blocking): We should add a comment explaining that this is to store a lock per wallet where the key is the wallet-id and the value is the `Lock` object.

File: src/lock.js [hunk 2]
```diff
@@ -68,4 +69,42 @@ class Lock {
...
+    this.globalLock = new Lock();
+  }
+
+  start(walletId) {
```
Comment (pedroferreira1) [inline src/lock.js:78]:
```suggestion
  _start(walletId) {
```

Comment (pedroferreira1) [reply src/lock.js:78]: suggestion (non-blocking): this method shouldn't be used outside of the class.

File: src/lock.js [hunk 3]
```diff
@@ -68,4 +69,42 @@ class Lock {
...
+
+  delete(walletId) {
+    if (this.wallets[walletId]) {
+      delete this.wallets[walletId];
+    }
+  }
+
+  get(walletId) {
+    if (!this.wallets[walletId]) {
+      return this.start(walletId);
```
Comment (pedroferreira1) [inline src/lock.js:93]: suggestion (non-blocking): remove `start` method. We can have the line `this.wallets[walletId] = new Lock();` here and remove the `start` method. What do you think?

Comment (r4mmer) [reply src/lock.js]: Done

File: src/lock.js (hunk 2 above)
Comment (r4mmer) [reply src/lock.js]: start method was removed

File: src/lock.js (hunk 1 above)
Comment (r4mmer) [reply src/lock.js:77]: Done

**Review by pedroferreira1 — APPROVED** [review]

**Review by pedroferreira1 — APPROVED** [review]

**Review by tuliomir — DISMISSED** [review]

File: src/controllers/index.controller.js [hunk 4]
```diff
@@ -304,12 +303,6 @@ async function pushTxHex(req, res) {
     return;
   }
 
-  const canStart = lock.lock(lockTypes.SEND_TX);
```
[the remark this answers is not in the document]
Comment (r4mmer) [reply src/controllers/index.controller.js:307]: Done

**Review by tuliomir — APPROVED** [review]

**Review by pedroferreira1 — APPROVED** [review]

**Pull request merged**